File: solutions/fsharp/robot-simulator/1/RobotSimulator.fs
module RobotSimulator

type Direction = 
    | North = 0 
    | East  = 1 
    | South = 2 
    | West  = 3

let offsetForDirection = Map.ofList [
    (Direction.North, (0, 1)); 
    (Direction.East, (1, 0)); 
    (Direction.South, (0, -1)); 
    (Direction.West, (-1, 0))
]

type Position = int * int
type Robot = { Direction: Direction; Position: Position }

let create direction position = { Direction = direction; Position = position }

let rotate increment direction = 
    let dirInt = (int direction + increment) % 4
    enum<Direction> dirInt

let turnLeft (robot: Robot) : Robot =
    { robot with Direction = rotate 3 robot.Direction }

let turnRight (robot: Robot) : Robot =
    { robot with Direction = rotate 1 robot.Direction }

let advance robot = 
    let x, y = robot.Position
    let dx, dy = offsetForDirection[robot.Direction]
    { robot with Position = (x + dx, y + dy)}

let rec move instructions robot = 
    match instructions with
    | "" -> robot
    | _ -> 
        let movedRobot = 
            match instructions[0] with
            | 'L' -> turnLeft robot
            | 'R' -> turnRight robot
            | 'A' -> advance robot
            | _ -> robot
        move instructions[1..] movedRobot

File: solutions/cpp/robot-simulator/1/robot_simulator.cpp
#include "robot_simulator.h"
#include <map>

namespace robot_simulator {

Robot::Robot(std::pair<int, int> position, Bearing bearing): position(position), bearing(bearing) {
    
}

void Robot::turn_right() {
    bearing = bearing == WEST ? NORTH : (Bearing) ((int) bearing + 1);
}

void Robot::turn_left() {
    bearing = bearing == NORTH ? WEST : (Bearing) ((int) bearing - 1);
}

void Robot::advance() {
    static const std::map<Bearing, std::pair<int, int>> offsets_for_bearing = {
        {NORTH, {0, 1}},
        {EAST, {1, 0}},
        {SOUTH, {0, -1}},
        {WEST, {-1, 0}}
    };

    position.first += offsets_for_bearing.at(bearing).first;    
    position.second += offsets_for_bearing.at(bearing).second;    
}    
    
void Robot::execute_sequence(const std::string& commands) {
    for (auto c: commands) {
        switch(c) {
            case 'L': 
                turn_left();
                break;
            case 'R':
                turn_right();
                break;
            case 'A':
                advance();
                break;
        }
    }
}

}  // namespace robot_simulator
